fix:: Implement graceful shutdown for Byzantine integration tests in …

shamb0 opened the pull request:
## This PR:

This PR addresses issue #3440 (https://github.com/EspressoSystems/HotShot/issues/3440) by refactoring the `SystemContextHandle` shutdown behavior. The changes implement a graceful shutdown process that works consistently across both `tokio` and `async-std` contexts.

## Key Changes

1. Introduced a new `shutdown_signal` field in `SystemContextHandle` to manage graceful shutdown for:
   - Event relay tasks (both send & receive) between consensus and network layers
   - Network message handler tasks

2. Refactored `EventTransformerState::add_network_tasks` to incorporate the new shutdown mechanism.

3. Modified `NetworkTaskRegistry::shutdown()` to use `join_all()` variant APIs instead of `abort()` variants, ensuring cleaner task exits.

## Validation

The changes have been tested on both tokio and async-std runtimes. Execution traces are available for review:

**tokio ::**

- [01-trace-tokio-byzantine-test-double-propose-vote.md](https://gist.github.com/shamb0/5cdb4496d2d9868c67c144fabd2ba03f)
- [02-trace-tokio-byzantine-test-multiple-bad-proposals.md](https://gist.github.com/shamb0/fc32dcb4d3609cf9115546d9cfbc4e8f)

**async-std ::**

- [03-trace-async-std-byzantine-test-double-propose-vote.md](https://gist.github.com/shamb0/8cf98204898528f05f27d9c86f2f82eb)
- [04-trace-async-std-byzantine-test-multiple-bad-proposals.md](https://gist.github.com/shamb0/00072c2472f151cb757902038f191406)

**This PR does not:**

**Key places to review:**

1. Implementation of the shutdown signal in `SystemContextHandle`
2. Integration of shutdown logic in event relay and network message handler tasks
3. Changes to `NetworkTaskRegistry::shutdown()` method

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EspressoSystems/HotShot?pullRequest=3535) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EspressoSystems/HotShot?pullRequest=3535) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EspressoSystems/HotShot?pullRequest=3535) it.</sub>

jparr721 commented on crates/hotshot/src/types/handle.rs:
```diff
@@ -170,14 +243,17 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES> + 'static> SystemContextHandl
         tracing::error!("Shutting down consensus!");
         self.consensus_registry.shutdown().await;
+
+        tracing::error!("=== Shutting down node :: {} done!!! ===", self.hotshot.id);
```
We probably don't need to double log the node shutdown.

jparr721 commented on crates/hotshot/src/types/handle.rs:
```diff
@@ -161,6 +230,10 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES> + 'static> SystemContextHandl
 
     /// Shut down the the inner hotshot and wait until all background threads are closed.
     pub async fn shut_down(&mut self) {
+        tracing::error!("=== Shutting down node :: {} ===", self.hotshot.id);
```
We probably don't need to double log the node shutdown.

jparr721 commented on crates/task/src/task.rs:
```diff
@@ -170,14 +170,11 @@ impl NetworkTaskRegistry {
-            #[cfg(async_executor_impl = "tokio")]
-            handle.abort();
-        }
+        let handles = std::mem::take(&mut self.handles);
```
Nice

jparr721 commented on crates/hotshot/src/tasks/mod.rs:
```diff
@@ -309,36 +326,57 @@ where
-                    let mut state = state_out.write().await;
+                futures::select! {
+                    _ = shutdown => {
+                        tracing::error!("Shutting down relay send task !!!");
```
We probably don't need all the exclamations here or in the below log.

jparr721 commented on crates/hotshot/src/tasks/mod.rs:
```diff
@@ -101,42 +102,57 @@ pub fn add_network_message_task<
-                            Err(e) => {
-                                tracing::error!("Failed to deserialize message: {}", e);
-                                continue;
+            if shutdown_signal.is_shutdown() {
+                tracing::error!("Shutting down network message task !!!");
+                break;
+            }
+            futures::select! {
+                    _ = shutdown => {
+                        tracing::error!("Shutting down network message task !!!");
+                        return;
+                    }
```
It looks like in this particular case, we have to handle the signal event twice, could we not just keep this in the future select?

jparr721 commented on crates/hotshot/src/types/handle.rs:
```diff
@@ -22,6 +29,65 @@ use tracing::instrument;
+/// This struct provides methods to initiate a shutdown, check the shutdown status,
+/// and wait for a shutdown signal.
+pub struct ShutdownSignal {
+    event: Arc<OtherEvent>,
+    is_shutdown: Arc<AtomicBool>,
```
Can we get doc comments here?

jparr721 commented on crates/hotshot/src/types/handle.rs:
```diff
@@ -1,12 +1,19 @@
 use async_lock::RwLock;
 #[cfg(async_executor_impl = "async-std")]
 use async_std::task::JoinHandle;
+use event_listener::Event as OtherEvent;
```
Can we use a different name for this? Perhaps `EventListenerEvent`, I'd even be okay with the specificity of `ShutdownSignalEvent` since that's all we're using it for.

jparr721 commented on crates/hotshot/src/types/handle.rs:
```diff
@@ -22,6 +29,65 @@ use tracing::instrument;
+///
+/// This struct provides methods to initiate a shutdown, check the shutdown status,
+/// and wait for a shutdown signal.
+pub struct ShutdownSignal {
```
I am hesitant about the introduction of a bespoke shutdown signal since we have something like this already in the codebase. Perhaps @ss-es can give context on the byzantine tests specifically, but it seems like we might be able to somehow re-use that behavior? Especially considering that this type is bolted onto the SYstemContextHandle anyway.

ss-es replied:
@shamb0 I think this looks good overall, though I agree with Jarred about the extra shutdown signal.

I think the preferred way to do this would be to listen on the `internal_event_stream` for a `HotShotEvent::Shutdown` (which is the shutdown signal for all tasks). I don't think the network message task has an internal event stream receiver, but you can pass it to the task state on creation in `tasks/mod.rs`.

ss-es replied:
Oh, nevermind -- I see you already made this change! I'll look through the code and add some comments

jparr721 reviewed (changes requested):
Hey there! Thanks for your contribution to the codebase. I have left some cursory comments. At a high level, this change seems good but, I have some overall concerns about the implementation in the context of the rest of the codebase.

I think that, mechanically, I don't mind this decision to use a shutdown signal, but I think that the tricky point here is we now need to arduously copy the same `futures::select` block in every single spot, and it increases the overall burden of replacing or modifying the behavior. 

Since the bulk of the change for the signal handler is in the `mod.rs` file, perhaps we can find a way to wrap this at a higher scope to avoid needless copy-pasted code?

Lastly and, most importantly, however, seems to be that we might be masking a deeper issue by having multiple shutdown handlers. We have all of our registries that are shutting down with some issues in Tokio preventing a graceful termination. It seems almost as though we're fixing the water damage without addressing the leaky roof, metaphorically speaking. Perhaps it's worth a deeper investigation.

shamb0 commented:
Hi @jparr721,

Thank you for your comments and for pointing out the potential duplicates. I have removed "ShutdownSignal" and reused the existing `EventDependency`. Additionally, Updated the "futures::select!" pattern in the Data Relay Task to align with the codebase's standard usage.

Please let me know if any further enhancements or corrections are needed.

bfish713 commented on crates/hotshot/src/tasks/mod.rs:
```diff
@@ -207,6 +221,42 @@ pub async fn add_consensus_tasks<TYPES: NodeType, I: NodeImplementation<TYPES>>(
+    // Activate the cloned internal event stream
+    let event_stream = handle.internal_event_stream.1.activate_cloned();
+
+    // Create an EventDependency that triggers on HotShotEvent::Shutdown
+    let event_dep = EventDependency::new(
+        event_stream,
+        Box::new(|e| matches!(e.as_ref(), HotShotEvent::Shutdown)),
+    );
+
+    // Create a future that completes when the EventDependency is satisfied
+    async move {
+        let _ = event_dep.completed().await;
+    }
+    .boxed()
```
EventDependency is basically just this while loop except it returns the event at the end so the `_` here would alwasy just be `HotShotEvent::Shutdown`   I'd also prefer just doing the loop here since we don't actually need Event returned

shamb0 replied:
Thanks @ss-es, @bfish713, Implemented a simple manual loop to check for `HotShotEvent::Shutdown`. No longer relying on `EventDependency`.

shamb0 commented:
Thanks @ss-es, Clippy warnings addressed. CI build should now pass. 
Ran cargo clippy checks with both Tokio and Async-std runners.

	```shell
		just async="tokio" cargo clippy --tests --examples
		just async="async-std" cargo clippy --tests --examples
	```

shamb0 commented:
Two quick points:

1. Disregard changes in these files:
   - crates/testing/tests/tests_1/test_with_failures_2.rs
   - crates/testing/tests/tests_1/vote_dependency_handle.rs

2. These files were unintentionally included when merging 
to upstream main. I'll review my git merge configuration on 
my dev machine to prevent this in the future.

jparr721 commented:
Thanks so much again for your great work!

shamb0 commented:
> Thanks so much again for your great work!

Thankyou @jparr721

ss-es commented:
It looks like there's some merge conflicts @shamb0 

(if you can't fix them, I'm happy to do it later tonight or tomorrow if you don't mind me pushing to the branch)

shamb0 commented:
> It looks like there's some merge conflicts @shamb0
> 
> (if you can't fix them, I'm happy to do it later tonight or tomorrow if you don't mind me pushing to the branch)

Thanks @ss-es: Branch conflicts have been resolved. The PR should now be ready for the Intake merger review.

ss-es commented:
merged -- thank you again for the PR!
